File: src/entities/track/api/library.ts
import { command } from "@/shared/api";
import type { ImportSummary, Track, TrackTagColor } from "../model/types.ts";

export const libraryApi = {
  getTracks: () => command<Track[]>("library_get_tracks"),
  getTrack: (trackId: string) => command<Track>("library_get_track", { trackId }),
  importFiles: (paths: string[]) => command<ImportSummary>("library_import_files", { paths }),
  importDirectory: (path: string) => command<ImportSummary>("library_import_directory", { path }),
  checkMissing: () => command<Track[]>("library_check_missing"),
  setFavorite: (trackId: string, favorite: boolean) => command<Track>("library_set_favorite", { trackId, favorite }),
  removeTrack: (trackId: string) => command<void>("library_remove_track", { trackId }),
  createTag: (trackId: string, label: string, color: TrackTagColor) => command<Track>("track_tag_create", { trackId, label, color }),
  deleteTag: (trackId: string, tagId: string) => command<Track>("track_tag_delete", { trackId, tagId }),
};

File: src-tauri/src/commands/library.rs
use crate::{
    dto::{ImportSummary, TrackDto},
    error::AppResult,
    library,
    state::AppState,
};
use std::path::PathBuf;
use tauri::State;

#[tauri::command]
pub async fn library_get_tracks(state: State<'_, AppState>) -> AppResult<Vec<TrackDto>> {
    library::all(&state.db).await
}
#[tauri::command]
pub async fn library_get_track(
    track_id: String,
    state: State<'_, AppState>,
) -> AppResult<TrackDto> {
    library::one(&state.db, &track_id).await
}
#[tauri::command]
pub async fn library_import_files(
    paths: Vec<String>,
    state: State<'_, AppState>,
) -> AppResult<ImportSummary> {
    Ok(library::import_paths(&state.db, paths.into_iter().map(PathBuf::from).collect()).await)
}
#[tauri::command]
pub async fn library_import_directory(
    path: String,
    state: State<'_, AppState>,
) -> AppResult<ImportSummary> {
    library::import_directory(&state.db, PathBuf::from(path)).await
}
#[tauri::command]
pub async fn library_check_missing(state: State<'_, AppState>) -> AppResult<Vec<TrackDto>> {
    library::check_missing(&state.db).await
}
#[tauri::command]
pub async fn library_set_favorite(
    track_id: String,
    favorite: bool,
    state: State<'_, AppState>,
) -> AppResult<TrackDto> {
    sqlx::query("UPDATE tracks SET is_favorite=?,updated_at=? WHERE id=?")
        .bind(favorite)
        .bind(chrono::Utc::now().timestamp_millis())
        .bind(&track_id)
        .execute(&state.db)
        .await?;
    library::one(&state.db, &track_id).await
}
#[tauri::command]
pub async fn library_remove_track(track_id: String, state: State<'_, AppState>) -> AppResult<()> {
    sqlx::query("DELETE FROM tracks WHERE id=?")
        .bind(track_id)
        .execute(&state.db)
        .await?;
    Ok(())
}
